open-metadata-implementation/access-services/data-engine/data-engine-client/src/main/java/org/odpi/openmetadata/accessservices/dataengine/client/DataEngineClient.java

```diff
@@ -24,7 +25,7 @@
 public interface DataEngineClient {
 
     /**
-     * Creates the process, with all the ports, schema types and corresponding relationships
+     * Create or update the process, with all the ports, schema types and corresponding relationships
```

Comment:
Would it make sense to have update semantics on all of these revised methods as well (perhaps again 2 methods for each: one that defaults to `REPLACE` and another that takes the update semantic explicitly)?

Reply:
I added the update semnatic just for port implementations and processes for now. It doesn't make sense to add it at SchemaType or PortAlias level, since I can not determine what is the old SchemaType/PortAlias without having the PortImplementation/Process in the payload.
For a process with UpdateSemnatic.REPLACE, it will remove all old Port Implementations & PortAliases and create all new ones based on the process payload.
For a PortImplementation with UpdateSemnatic.REPLACE, it will remove the old SchemaType(&attributes) and create the new one based on the port implementation payload.

Reply:
Just to check, when we say "replace" do we mean it literally does a full delete / purge of any existing pieces and a create (new) of everything?  Or is there some sort of determination of what's no longer in the new payload that was there previously, and only deleting those pieces + adding net-new ones that appear in the payload and are not found already in the environment?

Just thinking about the potential impact if we do a complete delete & complete new creation each time: that would create all new GUIDs for all of the underlying pieces, and potentially that would mean we'd lose any existing relationships against those underlying pieces?  My hunch is that while such an approach would be easier overall, it has greater potential for information being lost (?)

For example: DataStage sends through information about a new process and its assets. This all gets created via the Data Engine OMAS.  Then, somewhere else (even in Egeria itself) someone creates a new SemanticAssignment relationship to one of those pieces (either a process, one of its underlying SchemaTypes, or one of the assets that is read from / written to).  If we then have an update from DataStage with "replace" semantic, but actually the only piece that is different in the update is that there is one more asset that is written (say a log file), but we delete and purge _everything_ and then create _everything_ new, we will lose that SemanticAssignment relationship that was set in-between on one of the pre-existing assets (which actually hasn't changed)...

Reply:
Not everything will be deleted and recreated - only the entitites that are no longer in the payload will be removed. I'm determining what was changed bassed on the existing relationships. 
For example, for PortImlementations - DE is updating the ones from the payload, then retrieving all the PortImplementations for that process based on the ProcessPort relationship, then removing every port that has a relationship and is not in the payload. Only after removing the old ports I'm creating the new ProcessPort relationships.

If I take your example, then in the payload sent from DataStage proxy I will have all the old ports and schemas, plus an extra port implenentation with a schema type. In this case, nothing will be removed, the new entities will be added and the existing ones will be updated. 

I'm not sure about the performance though, haven't done enough tests yet. I tried to paralelize everything as much as possible, but for each entity from the payload now there is an extra call to retrieve it from the repository based on qualifiedName. If it is retrieved, then there is another call for updating that entity.

Reply:
Excellent!

I've tried to emphasise whenever possible the criticality of `find...` methods working as fast as possible on `qualifiedName` for precisely these sorts of reasons -- that they will be called MANY times, so even what appears to be a tiny optimisation will have a significant impact 😉